### ISSUE TITLE
`__pthread_tpp_change_priority` error in GenotypeGVCFs (GATK 4.0.0.0)

### ISSUE THREAD

**Issue by jrandall** [broadinstitute/gatk#4518]
About 3% of our GATK 4.0.0.0 GenotypeGVCFs runs (with a GenomicsDB as input) are failing with a `__pthread_tpp_change_priority` error and exiting with status -6.

The stderr of such a run ends like this:
```
2018-03-10T07:14:27.165578644Z GENOMICSDB_TIMER,GenomicsDB iterator next() timer,Wall-clock time(s),74.01474455399772,Cpu time(s),62.96424693700022
2018-03-10T07:14:27.168329699Z java: tpp.c:84: __pthread_tpp_change_priority: Assertion `new_prio == -1 || (new_prio >= fifo_min_prio && new_prio <= fifo_max_prio)' failed.
```

Stdout from the same run:
```
2018-03-09T13:13:41.095913747Z 13:13:41.095 INFO  NativeLibraryLoader - Loading libgkl_compression.so from jar:file:/gatk/gatk.jar!/com/intel/gkl/native/libgkl_compression.so
2018-03-09T13:13:41.329888610Z 13:13:41.327 INFO  GenotypeGVCFs - ------------------------------------------------------------
2018-03-09T13:13:41.329934964Z 13:13:41.327 INFO  GenotypeGVCFs - The Genome Analysis Toolkit (GATK) v4.0.0.0
2018-03-09T13:13:41.329942970Z 13:13:41.327 INFO  GenotypeGVCFs - For support and documentation go to https://software.broadinstitute.org/gatk/
2018-03-09T13:13:41.329952404Z 13:13:41.328 INFO  GenotypeGVCFs - Executing as root@localhost on Linux v4.4.0-112-generic amd64
2018-03-09T13:13:41.329960555Z 13:13:41.328 INFO  GenotypeGVCFs - Java runtime: OpenJDK 64-Bit Server VM v1.8.0_151-8u151-b12-1~deb9u1-b12
2018-03-09T13:13:41.329988865Z 13:13:41.328 INFO  GenotypeGVCFs - Start Date/Time: March 9, 2018 1:13:41 PM UTC
2018-03-09T13:13:41.329995417Z 13:13:41.328 INFO  GenotypeGVCFs - ------------------------------------------------------------
2018-03-09T13:13:41.330000910Z 13:13:41.328 INFO  GenotypeGVCFs - ------------------------------------------------------------
2018-03-09T13:13:41.330011002Z 13:13:41.328 INFO  GenotypeGVCFs - HTSJDK Version: 2.13.2
2018-03-09T13:13:41.330022980Z 13:13:41.328 INFO  GenotypeGVCFs - Picard Version: 2.17.2
2018-03-09T13:13:41.330030226Z 13:13:41.329 INFO  GenotypeGVCFs - HTSJDK Defaults.COMPRESSION_LEVEL : 1
2018-03-09T13:13:41.330036559Z 13:13:41.329 INFO  GenotypeGVCFs - HTSJDK Defaults.USE_ASYNC_IO_READ_FOR_SAMTOOLS : false
2018-03-09T13:13:41.330045071Z 13:13:41.329 INFO  GenotypeGVCFs - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_SAMTOOLS : true
2018-03-09T13:13:41.330051564Z 13:13:41.329 INFO  GenotypeGVCFs - HTSJDK Defaults.USE_ASYNC_IO_WRITE_FOR_TRIBBLE : false
2018-03-09T13:13:41.330068542Z 13:13:41.329 INFO  GenotypeGVCFs - Deflater: IntelDeflater
2018-03-09T13:13:41.330102470Z 13:13:41.329 INFO  GenotypeGVCFs - Inflater: IntelInflater
2018-03-09T13:13:41.330111084Z 13:13:41.329 INFO  GenotypeGVCFs - GCS max retries/reopens: 20
2018-03-09T13:13:41.330117286Z 13:13:41.329 INFO  GenotypeGVCFs - Using google-cloud-java patch 6d11bef1c81f885c26b2b56c8616b7a705171e4f from https://github.com/droazen/google-cloud-java/tree/dr_all_nio_fixes
2018-03-09T13:13:41.330127806Z 13:13:41.329 INFO  GenotypeGVCFs - Initializing engine
2018-03-09T13:13:44.528605497Z 13:13:44.528 INFO  GenotypeGVCFs - Done initializing engine
2018-03-09T13:13:45.237843760Z 13:13:45.235 INFO  ProgressMeter - Starting traversal
2018-03-09T13:13:45.237903383Z 13:13:45.235 INFO  ProgressMeter -        Current Locus  Elapsed Minutes    Variants Processed  Variants/Minute
2018-03-09T13:13:56.665869885Z 13:13:56.662 INFO  ProgressMeter -       chr13:82078938              0.2                 13000          68265.4
2018-03-09T13:14:07.517952300Z 13:14:07.517 INFO  ProgressMeter -       chr13:82096938              0.4                 31000          83475.5
2018-03-09T13:14:17.546110604Z 13:14:17.545 INFO  ProgressMeter -       chr13:82123938              0.5                 58000         107706.6
2018-03-09T13:14:28.760222694Z 13:14:28.759 INFO  ProgressMeter -       chr13:82144938              0.7                 79000         108905.4
2018-03-09T13:14:39.292149466Z 13:14:39.289 INFO  ProgressMeter -       chr13:82169938              0.9                104000         115440.1
2018-03-09T13:14:49.877851947Z 13:14:49.873 INFO  ProgressMeter -       chr13:82193938              1.1                128000         118815.6
2018-03-09T13:15:01.109839332Z 13:15:01.106 INFO  ProgressMeter -       chr13:82224938              1.3                159000         125741.4
2018-03-09T13:15:11.237868051Z 13:15:11.234 INFO  ProgressMeter -       chr13:82250938              1.4                185000         129071.3
2018-03-09T13:15:22.457899462Z 13:15:22.455 INFO  ProgressMeter -       chr13:82284938              1.6                219000         135157.4
2018-03-09T13:15:32.561846058Z 13:15:32.556 INFO  ProgressMeter -       chr13:82319938              1.8                254000         142003.9
2018-03-09T13:15:42.624917849Z 13:15:42.624 INFO  ProgressMeter -       chr13:82348938              2.0                283000         144647.3
2018-03-09T13:32:20.050317641Z 13:32:20.049 INFO  ProgressMeter -       chr13:82369938             18.6                304000          16361.5
2018-03-09T13:47:24.213333084Z 13:47:24.212 INFO  ProgressMeter -       chr13:82373938             33.6                308000           9153.2
2018-03-09T14:21:34.139657997Z 14:21:34.139 INFO  ProgressMeter -       chr13:82377938             67.8                312000           4600.7
2018-03-10T07:14:27.162870189Z 07:14:27.161 INFO  GenotypeGVCFs - Shutting down engine
```

**Comment by jrandall** [comment]
Repeating the GenotypeGVCFs run on exactly the same input data results in success, so it appears to be a nondeterministic error.

**Comment by droazen** [comment]
@kgururaj Could you and your team look into this one?

**Comment by kgururaj** [comment]
We aren't using any multi-threading inside  the GenomicsDB jar. If there is a way to reproduce this, we can take a look.

**Comment by lbergelson** [comment]
Here's some reports of similar errors:

https://sourceware.org/ml/libc-help/2008-05/msg00072.html
https://stackoverflow.com/questions/36221574/random-error-on-mutex-lock

It seems like most commonly this error is caused by forgetting to initialize a `pthread_mutexattr_t`

Although the second answer on this thread claims that it might be due to using the default mutex type instead of a recursive mutex.

https://stackoverflow.com/questions/21825291/threading-issues

GenomicsDB may not use pthreads directly, but both tileDB and htslib do, so I think it's worth looking into it.  I haven't found any instances of htslib or tileDb using an uninitialized `pthread_mutexattr_t` but it does look like tiledb uses default attributes for pthread_mutex_t instead of setting `PTHREAD_MUTEX_RECURSIVE`, so if that really is an issue maybe that's the problem?

**Comment by nalinigans** [comment]
Changed to initializing pthread mutex with `PTHREAD_MUTEX_ERRORCHECK` in a debug snapshot genomicsdb build 1.0.0-20181106.031500-10.  Created branch nalini_issue_4518 -https://github.com/broadinstitute/gatk/tree/nalini_issue_4518 using this genomicsdb.jar. 

The last call stack that could be gathered was -
```
(gdb) bt
#0  0x00007f1124858067 in __GI_raise (sig=sig@entry=6) at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00007f1124859448 in __GI_abort () at abort.c:89
#2  0x00007f1124851266 in __assert_fail_base (fmt=0x7f1124989f18 "%s%s%s:%u: %s%sAssertion `%s' failed.\n%n", 
    assertion=assertion@entry=0x7f112520df60 "new_prio == -1 || (new_prio >= __sched_fifo_min_prio && new_prio <= __sched_fifo_max_prio)", file=file@entry=0x7f112520df54 "tpp.c", line=line@entry=62, 
    function=function@entry=0x7f112520e030 <__PRETTY_FUNCTION__.8458> "__pthread_tpp_change_priority")
    at assert.c:92
#3  0x00007f1124851312 in __GI___assert_fail (
    assertion=assertion@entry=0x7f112520df60 "new_prio == -1 || (new_prio >= __sched_fifo_min_prio && new_prio <= __sched_fifo_max_prio)", file=file@entry=0x7f112520df54 "tpp.c", line=line@entry=62, 
    function=function@entry=0x7f112520e030 <__PRETTY_FUNCTION__.8458> "__pthread_tpp_change_priority")
    at assert.c:101
#4  0x00007f112520c5ef in __pthread_tpp_change_priority (previous_prio=previous_prio@entry=-1, 
    new_prio=new_prio@entry=6271) at tpp.c:60
#5  0x00007f1125201e5f in __pthread_mutex_lock_full (mutex=0x7f111cd7e340) at ../nptl/pthread_mutex_lock.c:453
#6  0x00007f10f4f83a2d in mutex_lock(pthread_mutex_t*) ()
   from /cromwell_root/tmp.034550cf/root/libtiledbgenomicsdb5206530589131345400.so
#7  0x00007f10f4efe245 in StorageManager::array_close(std::string const&) ()
   from /cromwell_root/tmp.034550cf/root/libtiledbgenomicsdb5206530589131345400.so
#8  0x00007f10f4efe8b7 in StorageManager::array_iterator_finalize(ArrayIterator*) ()
   from /cromwell_root/tmp.034550cf/root/libtiledbgenomicsdb5206530589131345400.so
#9  0x00007f10f4ef77af in tiledb_array_iterator_finalize ()
   from /cromwell_root/tmp.034550cf/root/libtiledbgenomicsdb5206530589131345400.so
#10 0x00007f10f4e44730 in GenomicsDBBCFGenerator::~GenomicsDBBCFGenerator() ()
   from /cromwell_root/tmp.034550cf/root/libtiledbgenomicsdb5206530589131345400.so
#11 0x00007f10f4e73491 in Java_com_intel_genomicsdb_GenomicsDBQueryStream_jniGenomicsDBClose ()
   from /cromwell_root/tmp.034550cf/root/libtiledbgenomicsdb5206530589131345400.so
```
@kgururaj, @droazen, would it be possible to use this branch and try gather more information to pinpoint the issue. Also, did not want to use `PTHREAD_MUTEX_RECURSIVE` while initializing the mutexes before figuring out the real issue.

**Comment by droazen** [comment]
@nalinigans Was this one fixed in 4.1?

**Comment by nalinigans** [comment]
4.1 has the improved reporting on system errors. Not sure if it fixes this issue. Would it be possible to include new logs by running 4.1 if you encounter the issue again?

**Comment by davidbenjamin** [comment]
@droazen Has this been fixed since 4.0.0.0?

**Comment by ldgauthier** [comment]
I don't remember an explicit fix, but I haven't seen seen that error in a long time and it hasn't come up in production.  We can reopen if it pops up again.
